Source/Core/AudioCommon/OpenALStream.cpp

```diff
@@ -277,74 +316,112 @@ void OpenALStream::SoundLoop()
     // many silence samples.  These do not need to be timestretched.
     if (rate > 0.10)
     {
-      soundTouch.setTempo(rate);
-      if (rate > 10)
+      if (use_soundtouch)
+      {
+        soundTouch.setTempo(rate);
+      }
+      else
+      {
+        alSourcef(uiSource, AL_PITCH, (ALfloat)rate);
+      }
+
+      if (rate > 10 && use_soundtouch)
       {
         soundTouch.clear();
       }
     }
 
-    unsigned int nSamples = soundTouch.receiveSamples(sampleBuffer, OAL_MAX_SAMPLES * numBuffers);
+    // the set latency of each buffer is not guaranteed when using soundtouch, but we need this or
+    // samples will accumulate causing high audio latency
+    if (use_soundtouch)
+    {
+      if (soundTouch.numSamples() < numSamples)
+        continue;
 
-    if (nSamples <= minSamples)
-      continue;
+      nSamples = soundTouch.receiveSamples(sampleBuffer.data(), soundTouch.numSamples());
+
+      // just to be sure that we have samples to play
+      if (nSamples < numSamples)
+        continue;
+    }
+    else
+    {
+      if (realtimeBufferSamples < numSamples)
+        continue;
+    }
 
     if (surround_capable)
     {
-      float dpl2[OAL_MAX_SAMPLES * OAL_MAX_BUFFERS * SURROUND_CHANNELS];
-      DPL2Decode(sampleBuffer, nSamples, dpl2);
+      // DPL2 accepts only floats
+      if (!use_soundtouch)
+        nSamples = realtimeBufferSamples;
+
+      std::vector<float> dpl2(nSamples * SURROUND_CHANNELS);
+
+      if (use_soundtouch)
+      {
+        DPL2Decode(sampleBuffer.data(), nSamples, dpl2.data());
+      }
+      else
+      {
+        std::vector<float> dest(nSamples * STEREO_CHANNELS);
+        for (u32 i = 0; i < nSamples * STEREO_CHANNELS; ++i)
+          dest.data()[i] = (float)realtimeBuffer.data()[i] / (1 << 15);
```

Comment:
Why are all these indexing operations using data? `std::vector` already provides an `operator[]` for direct indexing.